src/test/scala/tech/cryptonomic/conseil/tezos/DatabaseConversionsTest.scala

```diff
@@ -46,15 +48,95 @@ class DatabaseConversionsTest
       sut.extractBigDecimal(InvalidDecimal("1000A")) shouldBe 'empty
     }
 
+    "convert a tezos genesis block to a database row" in {
+      val converted = genesis.convertTo[Tables.BlocksRow]
+
+      val header = genesis.data.header
+      val CurrentVotes(expectedQuorum, proposal) = genesis.votes
+
+      converted should have(
+        'level (header.level),
+        'proto (header.proto),
+        'predecessor (header.predecessor.value),
+        'timestamp (java.sql.Timestamp.from(header.timestamp.toInstant)),
+        'validationPass (header.validation_pass),
+        'fitness (header.fitness.mkString(",")),
+        'context (Some(header.context)),
+        'signature (header.signature),
+        'protocol (genesis.data.protocol),
+        'chainId (genesis.data.chain_id),
+        'hash (genesis.data.hash.value),
+        'operationsHash (header.operations_hash),
+        'currentExpectedQuorum (expectedQuorum),
+        'activeProposal (proposal.map(_.id))
+      )
+
+      //no metadata expected
+      forAll(
+        converted.periodKind ::
+        converted.baker ::
+        converted.nonceHash ::
+        converted.consumedGas ::
+        converted.metaLevel ::
+        converted.metaLevelPosition ::
+        converted.metaCycle ::
+        converted.metaCyclePosition ::
+        converted.metaVotingPeriod ::
+        converted.metaVotingPeriodPosition ::
+        converted.expectedCommitment :: Nil) {
+        _ shouldBe 'empty
+      }
+
+    }
+
+    "convert a tezos block to a database row" in {
```

Comment:
I'd prefer to have a declaration and comparison in one place so I can see structure before and after conversion. Right now It looks like repeating production code here in tests.

Reply:
I feel your case, sure.
This is the result of having random data generated for the test. Which is why I would love to find time to rethink it in terms of generic properties to satisfy, instead of using a single manually made sample to check against.

This test is almost useless per se, as it duplicates code logic, but it might have its reason to be because
- we plan to try and use the Chimney library to do the conversion with less manual copying, in which case the test will be different code-wise (which is not such a strong case, nonetheless)
- I added the case because otherwise the test coverage percentage would treat this as "untested" code, and lower our rating... that's the price to pay if we want to have a public-facing decent coverage

Here too, I'm open to discussion

Reply:
Ok, so that's the reason I would love to have nice integration tests which are part of the coverage. Like in hexagonal architecture, where domain code can be tested with unit tests and all the adapters like postresql adapter and all conversions between domain and adapter's dtos are tested with integration tests (as part of the process and no in separate tests). 

Example:

> trait PostgresqlStorage {
>   def save(block: Block): Unit
>   def get(hash: String): Block
> }

Implementation of this adapter can be tested by saving some block and then receiving it back from DB with embedded postresql database run with a test suite. ``Block`` is the domain class and the conversion to database's dto will be tested by the way, as it is specific to a given database.

I think that this approach could solve both problems: testing implementation instead of behavior and coverage issues. What do you think?

Reply:
The actually like the current separation of Conversion logic (adapting the data) from the Storage abstraction (saving and fetching it).

It's easier an modular to both maintain and test in my opinion.

I'm with you as long as we talk about the DatabaseOperator (or anyway we want to call the Repository  module). I foresee the possibility of generalising the interface to the storage and provide parametric and separate implementations. That would also allow separate testing for the compliance of each and every implementation.

Reply:
Of course this leave open the fact that testing conversion of a randomly generated instance would always imply some form of duplication of logic... 
Also keep in mind that the testing adapter might become much simpler than the real one, if the conversion logic gets sophisticated.